YubiKit/YubiKit/Helpers/YKFTLVRecord.m

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018-2022 Yubico AB
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import "YKFTLVRecord.h"
+#import "YKFNSDataAdditions+Private.h"
+
+@interface YKFTLVRecord()
+@property (nonatomic, readwrite) YKFTLVTag tag;
+@property (nonatomic, readwrite) NSData *value;
+@end
+
+@interface NSData (NSData_ByteEncoding)
+    + (NSData *)ykf_encodeBytesBigSkippingZeros:(char*)bytes length:(int)length;
+@end
+
+@implementation YKFTLVRecord
+
++ (nullable instancetype)recordFromData:(NSData *_Nullable)data checkMatchingLength:(Boolean)checkMatchingLength bytesRead:(int*)bytesRead {
+    // tag
+    if (data.length == 0) { return nil; }
+    Byte *bytes = (Byte *)data.bytes;
+    int offset = 0;
+    int tag = bytes[offset++] & 0xFF;
+    if ((tag & 0x1F) == 0x1F) {
+        tag = (tag << 8) | (bytes[offset++] & 0xFF);
+        while ((tag & 0x80) == 0x80 && offset < data.length) {
+            tag = (tag << 8) | (bytes[offset++] & 0xFF);
+        }
+    }
+    
+    // length
+    int length = bytes[offset++];
+    if (length == 0x80) {
+        return nil;
+    } else if (length > 0x80) {
+        int lengthOfLength = length - 0x80;
+        length = 0;
+        if (data.length < offset + lengthOfLength) { return nil; }
+        for (int i = 0; i < lengthOfLength; i++) {
+            length = (length << 8) | (bytes[offset++] & 0xff);
+        }
+    }
+    // data
+    *bytesRead = offset + length;
```

Comment:
Is it a problem that `bytesRead` is set even if we return `nil` in one of the following 2 rows?

Reply:
It's only used in `sequenceOfRecordsFromData` and it's always set to 0 but to save the future me from embarrassment I'm changing it to set the proper value when failing to parse a record.